### PR TITLE
[SYCL][ESIMD] Add verifications that device has necessary aspects

### PR DESCRIPTION
**Pull request by vasilytric**
Provide verifications that device has necessary aspects to interact with `double` or `sycl::half` data types, if device doesn't has necessary aspects such data type will be skipped during runtime

### CODE REVIEW

File: SYCL/ESIMD/api/functional/common.hpp [hunk 1]
```diff
@@ -55,6 +55,24 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
 template <typename DataT>
 using shared_vector = std::vector<DataT, shared_allocator<DataT>>;
 
+template <typename T>
+constexpr bool should_run_test_with_current_datatype(sycl::device &&device) {
```
Comment (yuriykoch) [inline SYCL/ESIMD/api/functional/common.hpp]:
```suggestion
constexpr bool should_run_test_with_current_datatype(const sycl::device &device) {
```
We are not modifying it or passing along anyway, so why limit usability to the r-values only? What do you think?

File: SYCL/ESIMD/api/functional/common.hpp [hunk 2]
```diff
@@ -55,6 +55,24 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
...
+      return false;
+    }
+  } else if constexpr (std::is_same_v<T, double>) {
+    if (!device.has(sycl::aspect::fp64)) {
+      log::note(
+          "Device does not support double precision floating point operations");
+      return false;
+    }
+  }
+  return true;
```
Comment (yuriykoch) [inline SYCL/ESIMD/api/functional/common.hpp]:
```suggestion
  } else {
    // Suppress compilation warnings
    static_cast<void>(device);
    return true;
  }
```
BTW, what do you think about making the function `inline`?

File: SYCL/ESIMD/api/functional/ctors/ctor_array.hpp [hunk 3]
```diff
@@ -96,6 +96,9 @@ template <typename DataT, typename SizeT, typename TestCaseT> class run_test {
 
 public:
   bool operator()(sycl::queue &queue, const std::string &data_type) {
+    if (!should_run_test_with_current_datatype<DataT>(queue.get_device())) {
+      return true;
+    }
```
Comment (yuriykoch) [inline SYCL/ESIMD/api/functional/ctors/ctor_array.hpp]:
```suggestion
    if (!should_run_test_with<DataT>(queue.get_device())) {
      return true;
    }
```
As a suggestion.  Or maybe even
```suggestion
    if (should_skip_test_with<DataT>(queue.get_device())) {
      return true;
    }
```

File: SYCL/ESIMD/api/functional/common.hpp (hunk 1 above)
Comment (yuriykoch) [inline SYCL/ESIMD/api/functional/common.hpp]: Would you mind to provide a comment to the function, with usage example to have API documented?
For example, for some test case depending on `SrcT` and `DstT`.

**Comment by yuriykoch** [comment]
Thanks, Vasily

File: SYCL/ESIMD/api/functional/common.hpp (hunk 2 above)
Comment (vasilytric) [reply SYCL/ESIMD/api/functional/common.hpp]: this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/839/commits/1c069b5f4a52017db43d26010feca6a451457ef9#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R84-R86)

File: SYCL/ESIMD/api/functional/ctors/ctor_array.hpp (hunk 3 above)
Comment (vasilytric) [reply SYCL/ESIMD/api/functional/ctors/ctor_array.hpp]: the function was renamed [here](https://github.com/intel/llvm-test-suite/pull/839/commits/1c069b5f4a52017db43d26010feca6a451457ef9#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R71)

File: SYCL/ESIMD/api/functional/common.hpp (hunk 1 above)
Comment (vasilytric) [reply SYCL/ESIMD/api/functional/common.hpp]: comment was added [here](https://github.com/intel/llvm-test-suite/pull/839/commits/1c069b5f4a52017db43d26010feca6a451457ef9#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R58-R69)

Comment (vasilytric) [reply SYCL/ESIMD/api/functional/common.hpp]: this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/839/commits/1c069b5f4a52017db43d26010feca6a451457ef9#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0L59-R72)

**Comment by vasilytric** [comment]
@kbobrovs Would you mind to take a look?

**Review by v-klochkov — DISMISSED** [review]

File: SYCL/ESIMD/api/functional/common.hpp [hunk 4]
```diff
@@ -55,6 +55,40 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
 template <typename DataT>
 using shared_vector = std::vector<DataT, shared_allocator<DataT>>;
 
+// Provides verification that provided device has necessary aspects to interact
+// with current data type.
+//
+// Usage example:
+//
+// sycl::device device = queue.get_device();
+// if (should_skip_test_with(device)) {
```
Comment (v-klochkov) [inline SYCL/ESIMD/api/functional/common.hpp]: Minor comment: the example usage is incomplete. The call needs the data type.
In my opinion, the lines 60-69 should be simply removed.
Having the description at the lines 58-59 is sufficient.

Comment (kbobrovs) [reply SYCL/ESIMD/api/functional/common.hpp]: @vasilytric - please address review comments

Comment (vasilytric) [reply SYCL/ESIMD/api/functional/common.hpp]: @kbobrovs I've applied this comment in this commit: 4977d372526f159aac3737caa810e2c99f6dc359

**Comment by vasilytric** [comment]
@v-klochkov all comments has been applied, would you mind to merge this PR?

**Review by v-klochkov — APPROVED** [review]

**Pull request merged**